### PR TITLE
Increasing resource requests and limits for kube-apiserver to avoid frequent scaling

### LINKED ISSUE #255: Static/Formula-Based Shoot Cluster Control Plane Scaling
# Story
* As provider I want my shoot cluster control planes to scale automatically (within certain limits), so that my shoot clusters remain operational and respond fast.

# Motivation
Response time behaviour. Avoiding issues like #79. Generally being able to run small clusters cheaply and large clusters safely (also important for our own shoot'ed seed clusters).

This is the simpler form of #16 that we can do also without a working VPA (see below).

# Acceptance Criteria
- [ ] XS (2 nodes), S (10 nodes), M (50 nodes), L (100 nodes) and XL clusters (250 nodes) with actual workload run safely without downtime for a week (hopefully longer, but that's the KPI and then we will free the resources).
- [ ] Control plane shows acceptable utilisation (no overly large over-provisioning)
- [ ] Minimal control plane unavailability (that is actually the greatest problem/threat, that we, when scaling, cause downtimes, most critically on API server and ETCD)

# Implementation Proposal
Based on the number of nodes (upper limit as simplest solution) or, better, cores and memory of the same (best solution would be actual CPU/memory metrics, but that would duplicate HPA/VPA), "calculate" the control plane `requires`/`limits` (adapt whenever the worker pools in the shoot resource are changed):
- [ ] Use horizontal and vertical pod ~~auto~~scaling for all shoot cluster control plane components that can be scaled horizontally like the API server (also brings HA as a side factor; we need horizontal and vertical pod autoscaling to not end up having hundreds of tiny pods, i.e. (suggestion) let's go to 3 pods when the cluster reaches a certain size (below a certain load 1 is more optimal in regards to resource consumption), use vertical pod autoscaling then and only then add more pods if the cluster/control plane demands to scale beyond 1/2 the capacity of a seed cluster node).
- [ ] Use vertical pod ~~auto~~scaling for all shoot cluster control plane components that cannot be scaled horizontally like the scheduler, controller manager and most importantly ETCD.
- [ ] If we go for a more dynamic approach (but that would actually bring us to #16), make sure to implement some hysteresis logic to avoid too much pod rescheduling and control plane unavailability. See also https://github.com/kubernetes/kubernetes/issues/9043 (the community also wants to keep the pods/containers in-place, if only the resource limits change to avoid unavailability).

Background: While HPA seems safe to use for horizontal pod autoscaling, VPA gives a very poor impression (NPEs) and the v1.11 release notes even hint at larger problems and that the VPA will be revamped completely (https://kubernetes.io/docs/imported/release/notes for v1.11, search for VPA).
We can wait for it (basically #16), implement something ourselves or use simpler solutions (based on the number of nodes/available memory only that at least **makes our small clusters cheap and our large clusters possible**, in the case of seed clusters without hard-coded hacks in the Gardener).

The hard question is also to come up with "formula", see also [this guide](https://github.com/coreos/etcd/blob/master/Documentation/op-guide/hardware.md) for etcd resource need for different cluster sizes).

# Questions
* Can we use something like the [addon-resizer](https://github.com/kubernetes/autoscaler/tree/master/addon-resizer) that is also used by the [metrics-server](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/metrics-server.md); see also [this guide](https://github.com/coreos/etcd/blob/master/Documentation/op-guide/hardware.md) for etcd resource need for different cluster sizes).
* How will "burstable" pods (`requires` below `limits`) work together with priorities? Can we use this (with high priorities) for API server and ETCD, so that these pods get more resources dynamically without frequent rescheduling?

FYI: https://kubernetespodcast.com/episode/009-sre/
- etcd split into main+events
- GKE regular cluster has really only one master with one etcd
- Only regional GKE clusters have 3 masters/etcds
- etcd leader election (only required for these regional clusters) is troublesome, etcd in general is troublesome
- Key indicator of etcd is healthz endpoint also for Google
- One main reason for failure is disk i/o pressure that causes the API server to fail first and then the entire cluster (edited)

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?
- [ ] Unit tests are provided: Have you written automated unit tests?
- [ ] Integration tests are provided: Have you written automated integration tests?
- [ ] Minimum API exposure: If you have added/changed public API, was it really necessary/is it minimal?
- [ ] Operations guide: Have you updated the operations guide about ops-relevant changes?
- [ ] User documentation: Have you updated the READMEs/docs/how-tos about user-relevant changes?

### PR DESCRIPTION
**Pull request by ggaurav10**
**What this PR does / why we need it**:
Currently kube-apiserver scales up when a small cluster (mostly less than 10 nodes) is booting up mainly because of the load induced by core kubernetes components coming up, and small request values applied for the CPU. Increasing this value will result in one API server in fresh clusters.
Also, increasing the `limits` and `requests` reduces need for frequent scaling.
Changed `horizontal-pod-autoscaler-tolerance` to 0.2 from default 0.1 to provide wider bandwidth to hpa, so that effect of load, fluctuating around the threshold, can be smoothened.

**Which issue(s) this PR fixes**:
Partially addresses #255 

**Special notes for your reviewer**:

**Release note**:

```noteworthy user
The kube-apiservers do now get higher CPU/memory limits to avoid frequent and unneeded up-/down-scaling.
```
```noteworthy operator
The HPA tolerance flag has been changed to 0.2 for the shooted Seed clusters to provide wider bandwidth to the HPA (effect of load, fluctuating around the threshold, can be smoothened).
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Thank's, we'll first consider #399 as it changed similar places inside the code. You'll have to rebase after this is merged, will ping you.

File: charts/seed-controlplane/charts/kube-controller-manager/values.yaml [hunk 1]
```diff
@@ -22,3 +22,4 @@ resources:
 horizontalPodAutoscaler:
   downscaleDelay: 15m
   upscaleDelay: 1m
+  tolerance: 0.2
```
Comment (amshuman-kr) [inline charts/seed-controlplane/charts/kube-controller-manager/values.yaml]: Perhaps we can keep the default `tolerance` in the values.yaml to `0.1` and set it to `0.2` only for seeds here -> https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/controlplane.go#L316-L319?

Comment (amshuman-kr) [reply charts/seed-controlplane/charts/kube-controller-manager/values.yaml]: Also, do we need to tweak CoreDNS' limits because now `tolerance` is `0.2`?

**Comment by amshuman-kr** [comment]
> we'll first consider #399 as it changed similar places inside the code.

@rfranzke Are you sure you don't want to merge this PR before #399? I could rebase #399 later as well.

File: charts/seed-controlplane/charts/kube-controller-manager/values.yaml (hunk 1 above)
Comment (ggaurav10) [reply charts/seed-controlplane/charts/kube-controller-manager/values.yaml]: Yes, will keep the tolerance for shoot to 0.1
Thanks for suggestion.

For CoreDNS:
Current requested: 100m, Limit: 200m
with 0.2 tolerance and 80% threshold, coreDNS will be scaled at 96m CPU usage.
So tweaking limits should not be required. WDYT?

Comment (amshuman-kr) [reply charts/seed-controlplane/charts/kube-controller-manager/values.yaml]: Cool. Thanks!

**Review by rfranzke — APPROVED** [review]
LGTM, thank you.

**Pull request merged**